docs/advanced-performance-analysis.md

```diff
@@ -0,0 +1,226 @@
+
+One strategy is to run
+
+    perf list | less
+
+and in there, search for "stall" (using `/stall`) and "penalty". This will yield some results of available performance counters that allow you to identify which particular stall it may be. It's worth to know that you can specify a bunch of them in one command line (comma separated). Figuring out which ones to choose is really an exercise in brute force with a little educated guessing, there aren't _that many_ such counters so it won't take you too long.  In the motivating example, the culprit turned out to be `dsb2mite_switches.penalty_cycles`
+
+```
+ Performance counter stats for './fast circular_buffer_into_byte_buffer_1kb':
+
+           9943811      dsb2mite_switches.penalty_cycles                                   
+
+       0.408032273 seconds time elapsed
+```
+
+vs
+
+```
+ Performance counter stats for './slow circular_buffer_into_byte_buffer_1kb':
+
+         229285609      dsb2mite_switches.penalty_cycles                                   
+
+       0.511033898 seconds time elapsed
+```
+
+The names of these stalls/penalties can be quite opaque and understanding what exactly is going on can be rather hard without a pretty good understanding of how a particular CPU micro-architecture works. Nevertheless, being able to pinpoint why a slowdown happens can be very important as it may give information on what to change in the code.
+
+[Feel free to skip this] In case you're wondering what `dsb2mite_switches.penalty_cycles` is, a web search revealed a nicely written [blogpost](https://easyperf.net/blog/2018/01/18/Code_alignment_issues) which discusses something very similar. Bottom line is that modern Intel CPUs have a MITE (Micro-operation Translation Engine) which seems to read the instruction stream in 16 byte chunks. So if the alignment of your code is changed (for example by adding/removing an unrelated function to/from your binary) you may incur a penalty if your instructions aren't aligned by 16 bytes (remember, Intel instructions are variable length). This does sound scary (and for some code it is) but the good news is that this should _very rarely_ matter outside of micro-benchmarks. If your whole run time is doing exactly what that micro-benchmark is doing, you should probably care. But most real-world code does much more and more diverse work than some arbitrary micro-benchmarks. So in this particular example we can safely ignore the "regression" which was just unlucky placement of the hot function in the binary. Nevertheless it was important to understand why the regression happened so we could safely decide it does not matter in this instance.
```

Comment:
Do we really want these square brackets?

Reply:
no